src/coreclr/jit/fginline.cpp

```diff
@@ -586,8 +586,70 @@ class SubstitutePlaceholdersAndDevirtualizeWalker : public GenTreeVisitor<Substi
                 }
 
                 CORINFO_CONTEXT_HANDLE contextInput = context;
+                context                             = nullptr;
                 m_compiler->impDevirtualizeCall(call, nullptr, &method, &methodFlags, &contextInput, &context,
                                                 isLateDevirtualization, explicitTailCall);
+                if (context != nullptr && !m_compiler->compCurBB->HasFlag(BBF_INTERNAL))
+                {
+                    CORINFO_METHOD_INFO methInfo;
+                    if (m_compiler->info.compCompHnd->getMethodInfo(call->gtCallMethHnd, &methInfo, context) &&
+                        genActualType(JITtype2varType(methInfo.args.retType)) == genActualType(call->TypeGet()))
```

Comment:
Why are we checking the return type here?

Reply:
I was hitting a case where the type can mismatch and leading to an assertion later.

Reply:
Okay it turns out to be retbuf. I'm figuring out a fix for this.

Reply:
Resolved.